doc/database/multi_namespace_db_instances.md

```diff
@@ -0,0 +1,495 @@
+# Support redis databases in multiple namespaces
+
+## Design
+
+The existing Multi-DB design approach needs to be extended to take care of cases where a sonic device have multiple ASICs (NPUs). Each NPU will have a "namespaces" exclusively created and services like database, swss, syncd etc running in that namespace. 
+
+## The Multi NPU/namespace architecture 
+
+In the Multi NPU devices, the services could be broadly classified into
+
+	* Global services like database, snmp, pmon, telemetry running in the dockers running in the linux "host" network namespace.
+	
+	* NPU specific services like database, swss, syncd, bgp, teamd, lldp etc which runs in a separate "namespace" created. 
+	  The method used currently to seggregate services per NPU is "linux network namespace" and there is a one-to-one 
+	  mapping between the number of NPUs and linux network namesapces created.
+
+This approach is backward compatible to the Sonic devices with single NPU. The "globalDB" service could be considered equivalent to the database docker started in the linux "host" network namespace in single NPU device. We will discuss more on the design details of database docker service in this document.
+
+The database docker started in the linux "host" namespace can be called the "global DB" service. The redis databases available here (decided by contents of database_config.json) would be APPL_DB, CONFIG_DB used to store the system wide attributes like AAA, syslog, ASIC to interface name mapping etc. 
+
+There are also database docker started per NPU in a separate "namespace". The redis databases there will include all the DB's viz. APPL_DB, CONFIG_DB, ASIC_DB, COUNTERS_DB etc and is used to store the interface/counters/state etc specific to the interfaces/features on that NPU.
+
+There will be a config_db.json file per database service, it will be named "/etc/sonic/config_db.json" for the "global DB" and /etc/sonic/config_db{NS}.json for the database service in the {NS} namespace.
+
+## Enhancements to Multi-DB design to support Multiple namespaces
+
+The file database_config.json contains the startup configuration which dictates the redis server host/port/unix_socket configurations + various available databases ( eg: APP_DB, CONFIG_DB ) etc. With the introduction of multiple namespaces, there is a need for separate database_config.json per namespace. This config file is created in the "working redis directories" during the database docker service startup. 
+
+The database docker service started in the linux network namespace {NS} will use the "/var/run/redis{NS}" as the "working redis directory" to create the various files like redis.sock, sonic-db/database_config.json etc. For "globalDB" database service it would be at "/var/run/redis".
```

Comment:
the database docker service does not have to be started in the linux network namespace. suggest to change to "database service" for a NPU linux namespace should use /var/run/redis{NS{. as the working directory.

Reply:
Could you clarify "the database docker service does not have to be started in the linux network namespace" ?
If not in namespace, how to specify one database container instance?

For example, currently we are using `sonic-db-cli STATE_DB hset`, what will be new command link like?

Reply:
The database docker service needs to be started in some network namespace. It could be a new "net_ns" which we create, or the linux host network namespace. 

A new parameter is added to  sonic-db-cli, 
sonic-db-cli APPL_DB **-namespace <ns>** HGET VLAN_TABLE:Vlan10 mtu

if we don't specify the namespace parameter, it executes the redis command in the DB instance in the database docker running that namespace where we executed the sonic-db-cli command. 

Do let me know if we need more discussion on this.